README.md
# MMM-Carousel w/ Slide Navigation
> This is an extension to the [MagicMirror](https://github.com/MichMich/MagicMirror) project, allowing the modules to be displayed in a rotating carousel instead of displaying all of them at once.
> This version of the module was forked from [barnabycolby's MMM-Carousel](https://github.com/barnabycolby/MMM-Carousel).
>  
>    There are three modes available:
* `'global'` - All modules not cited in the `ignoreModules` config are rotated, displaying only one at a time for the duration of `transitionInterval`.  This is particularly useful on small screens where there may not be enough space to display several components at once. 
* `'positional'` - Modules are grouped by `position` setting and rotated within a position except for modules listed in that position's `ignoreModules`, an `overrideTransitionInterval` can also be set to rotated different position at different speeds.
* `'slides'` - groups of modules can be assigned to be displayed at the same time (regardless of `position`), an unlimited number of these "slide" groups can be set up. Module positions and appearances can also be changed for each slide.
* ***New:*** Slide Indicators and Navigation Buttons can be enabled (see config settings)

    ![](https://raw.githubusercontent.com/shbatm/MMM-Carousel/wip/img/slide.jpg)

    ![](https://raw.githubusercontent.com/shbatm/MMM-Carousel/wip/img/navbar.jpg)
    
    Screencast (from @AgilesChroms) : https://www.youtube.com/watch?v=2WTgQLaAMQk

* ***New:*** Modules can be moved to different positions and CSS classes applied to them for each slide.
* ***New:*** Multiple instances of a module can be used on different slides.
* ***New:*** Integration with [MMM-KeyBindings](https://github.com/shbatm/MMM-KeyBindings) for keyboard and bluetooth remote navigation.

## Installation
Run these commands at the root of your magic mirror install.

```shell
cd modules
git clone https://github.com/shbatm/MMM-Carousel
```

## Using the module
To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-Carousel',
            position: 'bottom_bar',  // Only required to show navigation
            config: {
                // See below for configurable options
            }
        }
    ]
}
```

Note that a `position` setting is required only if you want to show the page navigation icons and buttons.

### Configuration options
The following properties can be configured:

<table width="100%">
	<!-- why, markdown... -->
	<thead>
		<tr>
			<th>Option</th>
			<th width="100%">Description</th>
		</tr>
	<thead>
	<tbody>
		<tr>
			<td><code>mode</code></td>
			<td><ul><li><code>'global'</code> - All modules not cited in the <code>ignoreModules</code> config are rotated, displaying only one at a time for the duration of <code>transitionInterval</code>.  This is particularly useful on small screens where there may not be enough space to display several components at once. 
                </li><li><code>'positional'</code> - Modules are grouped by <code>position</code> setting and rotated within a position except for modules listed in that position's <code>ignoreModules</code>, an <code>overrideTransitionInterval</code> can also be set to rotated different position at different speeds.
                </li><li><code>'slides'</code> - groups of modules can be assigned to be displayed at the same time (regardless of <code>position</code>), an unlimited number of these "slide" groups can be set up.</li></ul>
				<br> <br> This value is <b>OPTIONAL</b>
				<br><b>Possible values:</b> <code>'global'</code> or <code>'positional'</code> or <code>'slides'</code>
                <br><b>Default value:</b> <code>'global'</code>
			</td>
		</tr>
		<tr>
			<td><code>transitionInterval</code></td>
			<td>The number of milliseconds to display each module for.
				<br> <br> This value is <b>OPTIONAL</b>
				<br><b>Possible values:</b> Any valid <code>int</code>, passing 0 with <code>mode:"slides"</code> will disable the timer for manual navigation.
                <br><b>Default value:</b> <code>10000</code>
			</td>
		</tr>
        <tr>
            <td><code>slideTransitionSpeed</code></td>
            <td>The speed in milliseconds to fade in each module.
                <br> <br> This value is <b>OPTIONAL</b>
                <br><b>Possible values:</b> Any valid <code>int</code>
                <br><b>Default value:</b> <code>1500</code> (ms)
            </td>
        </tr>
		<tr>
			<td><code>ignoreModules</code></td>
			<td>A list of module names whom should not be considered as part of the carousel. For example, the `alert` module should be able to display a notification at any time, by ignoring it we can prevent the plugin from hiding any notifications. <b> NOTE: is only used in <code>'global'</code> and <code>'slides'</code> modes.  Ignored modules in <code>'slides'</code> mode are shown on every slide.</b>.
				<br> <br> This value is <b>OPTIONAL</b>
				<br><b>Possible values:</b> <code>String array</code>
                <br><b>Default value:</b> <code>[]</code>
			</td>
		</tr>
		<tr>
			<td>
			    <code>top_bar</code>
			    <br><code>top_left</code>
			    <br><code>top_center</code>
			    <br><code>top_right</code>
			    <br><code>upper_third</code>
			    <br><code>middle_center</code>
			    <br><code>lower_third</code>
			    <br><code>bottom_left</code>
			    <br><code>bottom_center</code>
			    <br><code>bottom_right</code>
			    <br><code>bottom_bar</code>
			</td>
			<td>Determines if this position should be rotated and which modules in this position should be ignored.  <b>NOTE: is only used when <code>mode</code> is <code>'positional'</code> otherwise ignored</b>.
				<br> <br> This value is <b>OPTIONAL</b>
				<br><b>Possible values:</b> Object with keys;
				<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; <code>enabled</code>, a boolean to rotate this position or not,
				<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; <code>ignoredModules</code>, a <code>String array</code> of modules names to ignore.
				<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; <code>overrideTransitionInterval</code>, a <code>int</code> a transition time for this position only.
                <br><b>Default value:</b> <code>{enabled: false, ignoreModules: [], overrideTransitionInterval: 10000}</code>
			</td>
		</tr>
		<tr>
        			<td><code>slides</code></td>
        			<td>See Examples below. The slides will be rotated as a complete set using the <code>transitionInterval</code> setting.  Ingnored modules (<code>ignoreModules</code>) will be diplayed on all slides.
        				<br> <br> This value is <b>OPTIONAL</b>
        				<br><b>Possible values:</b> <code>array or Object of String/Object arrays</code> <a href="#advanced-slides">(see below)</a>
                        <br><b>Default value:</b> <code>[[]]</code>
        			</td>
        		</tr>
        <tr>
            <td><code>showPageIndicators</code></td>
            <td>A <code>boolean true or false</code> to show or hide the page indication 'bubbles' at the bottom of the screen. Default is <code>true</code>.
            </td>
        </tr>
        <tr>
            <td><code>showPageControls</code></td>
            <td>A <code>boolean true or false</code> to show or hide the next and previous page buttons. Buttons will only appear when hovered near or over. They can be clicked when not visible. Default is <code>true</code>.
            </td>
        </tr>
        <tr>
            <td><code>showPageTitles</code></td>
            <td>Valid options: <code>false</code> (default), <code>'top'</code>, or <code>'bottom'</code> to show the page titles at the top or bottom. The page titles are derived from the <a href="#advanced-slides">Slides config</a> provided. If you want to use page titles, it is recommended you use an Object with named slides in your config; if you use an array, you will just see the slide number ("0", "1", etc.).
            </td>
        </tr>
	<tr>
		<td><code>keyBindings: {</code><br>
            &nbsp;&nbsp;<code>enabled: true,</code><br>
            &nbsp;&nbsp;<code>mode: "DEFAULT",</code><br>
&nbsp;&nbsp;<code>map: {</code><br>
&nbsp;&nbsp;&nbsp;&nbsp;<code>NextSlide:"ArrowRight",</code><br>
&nbsp;&nbsp;&nbsp;&nbsp;<code>PrevSlide:"ArrowLeft",</code><br>
&nbsp;&nbsp;&nbsp;&nbsp;<code>Slide0:"Home"</code><br>
&nbsp;&nbsp;<code>}</code><br>
<code>}</code></td>
		<td>Key bindings to use for navigation with <A href="https://github.com/shbatm/MMM-KeyBindings">MMM-KeyBindings</A> plugin. The values are the KeyNames to respond to from the <code>"KEYPRESS"</code> events generated in MMM-KeyBindings.<br>Note: any <code>Slide##</code> can be assigned to jump to a specific slide.</td>
	</tr>
        <tr>
        <td><code>keyBindings.mode</code></td>
        <td>Mode Keyword for responding to key press events sent from <A href="https://github.com/shbatm/MMM-KeyBindings">MMM-KeyBindings</A>. Default: "DEFAULT" which repsonds to any key press when no other module has taken focus or changed the keyword.</td>
    </tr>
	</tbody>
</table>

#### Example - Global Carousel
```javascript
var config = {
    modules: [
        {
            module: 'MMM-Carousel',
            config: {
                transitionInterval: 10000,
                ignoreModules: ['clock'],
                mode: 'global'
            }
        }
    ]
}
```
#### Example - Positional Carousel
```javascript
var config = {
    modules: [
        {
            module: 'MMM-Carousel',
            config: {
                transitionInterval: 10000,
                ignoreModules: [],
                mode: 'positional',
                top_left: {enabled: true, ignoreModules: [], overrideTransitionInterval: 15000},
                top_right: {enabled: true, ignoreModules: ['currentweather']}
            }
        }
    ]
}
```
#### Example - Slides Carousel
```javascript
var config = {
    modules: [
        {
            module: 'MMM-Carousel',
            position: 'bottom_bar', // Required only for navigation controls
            config: {
                transitionInterval: 10000,
                showPageIndicators: true,
                showPageControls: true,
                ignoreModules: ['clock', 'alert'],
                mode: 'slides',
                slides: {
                    "Main": ['calendar', 'compliments', 'currentweather'],
                    "Slide 2": ['weatherforecast', 'MMM-Trello', 'planetrise', 'newsfeed'],
                    "Slide 3": ['MMM-fitbit']
                }
            }
        }
    ]
}
```
#### Example - Slides Carousel w/ [MMM-KeyBindings](https://github.com/shbatm/MMM-KeyBindings)
```javascript
var config = {
    modules: [
        {
            module: 'MMM-Carousel',
            position: 'bottom_bar', // Required only for navigation controls
            config: {
                transitionInterval: 10000,
                ignoreModules: ['clock', 'alert'],
                mode: 'slides',
                showPageIndicators: true,
                showPageControls: true,
                slides: {
                    main: ['calendar', 'compliments', 'currentweather'],

CHANGELOG.md
## [0.2.7] - Changed CarouselId lookup for advanced use cases

* Changed CarouselId match to an indexOf match to allow different configurations such as "all|2" to show on two different slide configurations, one with "all" and one with "2".
* Added Page Titles -- new config option `showPageTitles` is detailed in README.

## <a name="0_2_6"></a>[0.2.6] - Named Slide Support & Update for new MMM-KeyBindings

* Added support for named slides.
* Updated Key Handling for new MMM-KeyBindings methods.
* Broadcast API request to new MMM-Remote-Control API.

## <a name="0_2_5"></a>[0.2.5] - Added direct controls via Notification.

* Added control from Module Notification (See [README](README.md#Navigation-from-other-modules)).
* Removed unneccessary and unused tests and Grunt file.

## <a name="0_2_4"></a>[0.2.4] - Module Fix for bug in MichMich/MagicMirror#1140

* Uses 'MODULE_DOM_CREATED' notification instead of 'DOM_OBJECTS_CREATED'

## <a name="0_2_4"></a>[0.2.4] - Added slideTranistionSpeed Parameter per #4

* Added a parameter for slide transition speed. Default is 1500ms.


## <a name="0_2_3"></a>[0.2.3] - Fix for #1 - Fix timed transition for slides

* Bug fix for Issue #1 -- Correct issue where setting transitionInterval for a "slides" configuration doesn't actually cause the slides to transition automatically.

## <a name="0_2_2"></a>[0.2.2] - Added support for MMM-KeyBindings Instances

* Added support for using MMM-KeyBindings control with multiple instances of the MagicMirror open.  For example, you can use a remote on the screen attached to the Raspberry Pi (SERVER) and use a keyboard on another computer with the mirror running in a browswer (LOCAL) to independently change slides on the respective screens.


## <a name="0_2_1"></a>[0.2.1] - Multiple instances of a module

Changes:

* Added the ability to handle mutiple instances of a module. To use, add a `carouselId: "uniqueString"` to each modules' `config` section:
```
        {
            module: "clock",
            position: "top_left",
            config: {
				carouselId: "1",
                displayType: "both"
            }
        },
        {
            module: "clock",
            position: "top_left",
            config: {
				carouselId: 2,
			}
        },
        {
            module: 'MMM-Carousel',
            position: "bottom_bar",
            config: {
                mode: 'slides',
                slides: [
                    [   {name:'clock', classes:'zoom200', position:"top_left", carouselId: "1"} ],
                    [   {name:'clock', classes:'', position:"top_left", carouselId: 2},  ]]
            }
        },
```

## [0.2.0] - Added manual-only slides and per-slide positions & classes

Changes:

* Added the ability to pass an object with detail for a module on a per slide basis.  Passing a config similar to the following shows a large clock on the first slide and then a small clock and additional modules on the second.
```
    mode: 'slides',
    slides: [
        [   {name:'clock', classes:'zoom200', position:"middle_center"} ],
        [   {name:'clock', classes:'', position:"top_left"}, 
            {name:'calendar', position:'top_left'}, 
            'MMM-WunderGround', 
            'newsfeed'
        ]
```
* Made use of `zoom` classes introduced in [0.1.1]: using the method above, supported zooms are 070%, 080%, 090%, 125%, 150%, 175%, and 200%. Pass `classes:''` for 100%.
* Added support for indefinite slides -- set `transitionInterval: 0` for manual-transition only slides
* Added KeyPress events to go to specific slide index. (e.g. `Slide0: "Home"` in the KeyBindings would jump to the first slide when the home key is pushed). Works with any number of slides in the format `Slide#: "KeyName"` where # is the 0-based index of the slide.

## [0.1.1] - Added manual transitions

Changes:

* Added manual transitions between slides
* Added optional Pagination indicators at the bottom of the screen to show what slide is currently showing.
* Added hidden next/previous page controls that show on hover in each lower corner

## [0.1.0] - First public release

First public release


MMM-Carousel.js
/*global Module, MM, setInterval */
/* jshint esversion:6 */
Module.register('MMM-Carousel', {
    defaults: {
        transitionInterval: 10000,
        slideTransitionSpeed: 1500,
        ignoreModules: [],
        mode: 'global', //global || positional || slides
        top_bar: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        top_left: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        top_center: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        top_right: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        upper_third: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        middle_center: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        lower_third: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        bottom_left: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        bottom_center: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        bottom_right: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        bottom_bar: { enabled: false, ignoreModules: [], overrideTransitionInterval: 10000 },
        slides: [
            []
        ],
        showPageIndicators: true,
        showPageControls: true,
        showPageTitles: false,
        // MMM-KeyBindings mapping.
        keyBindings: {
            enabled: true
        }
    },

    keyBindings: {
        mode: "DEFAULT",
        map: { NextSlide: "ArrowRight", PrevSlide: "ArrowLeft", Slide0: "Home" }
    },

    requiresVersion: "2.3.0", // Uses 'MODULE_DOM_CREATED' notification instead of 'DOM_OBJECTS_CREATED'

    start: function() {

    },

    validKeyPress: function(kp) {
        if (kp.keyName === this.keyHandler.config.map.NextSlide) {
            this.manualTransition(undefined, 1);
            this.restartTimer();
        } else if (kp.keyName === this.keyHandler.config.map.PrevSlide) {
            this.manualTransition(undefined, -1);
            this.restartTimer();
        } else if (this.keyHandler.reverseMap[kp.keyName].startsWith("Slide")) {
            var goToSlide = this.keyHandler.reverseMap[kp.keyName].match(/Slide([0-9]+)/i);
            console.log((typeof goToSlide[1]) + " " + goToSlide[1]);
            if (typeof parseInt(goToSlide[1]) === "number") {
                this.manualTransition(parseInt(goToSlide[1]));
                this.restartTimer();
            }
        }
    },

    notificationReceived: function(notification, payload, sender) {
        var position, positions = ['top_bar', 'bottom_bar', 'top_left', 'bottom_left', 'top_center', 'bottom_center', 'top_right', 'bottom_right', 'upper_third', 'middle_center', 'lower_third'];
        if (notification === 'MODULE_DOM_CREATED') {
            // Register Key Handler
            if (this.config.keyBindings.enabled &&
                MM.getModules().filter(kb => kb.name === "MMM-KeyBindings").length > 0) {
                this.keyBindings = Object.assign({}, this.keyBindings, this.config.keyBindings);
                KeyHandler.register(this.name, {
                    validKeyPress: (kp) => {
                        this.validKeyPress(kp); // Your Key Press Function
                    }
                });
                this.keyHandler = KeyHandler.create(this.name, this.keyBindings);
            }

            // Initially, all modules are hidden except the first and any ignored modules
            // We start by getting a list of all of the modules in the transition cycle
            if ((this.config.mode === 'global') || (this.config.mode === 'slides')) {
                this.setUpTransitionTimers(null);
            } else {
                for (position = 0; position < positions.length; position += 1) {
                    if (this.config[positions[position]].enabled === true) {
                        this.setUpTransitionTimers(positions[position]);
                    }
                }
            }

            let api = {
                module: "MMM-Carousel",
                path: "carousel",
                actions: {
                    next: {
                        notification: "CAROUSEL_NEXT",
                        prettyName: "Next Slide"
                    },
                    previous: {
                        notification: "CAROUSEL_PREVIOUS",
                        prettyName: "Previous Slide"
                    },
                }
            };
            if (this.config.mode === 'slides') {
                Object.keys(this.config.slides).forEach(s => {
                    api.actions[(s.replace(/\s/g, '').toLowerCase())] = {
                        notification: "CAROUSEL_GOTO",
                        payload: { slide: s },
                        prettyName: "Go To Slide " + s
                    };
                });
            }
            this.sendNotification("REGISTER_API", api);
        }

        if (this.keyHandler && this.keyHandler.validate(notification, payload)) { return; }

        if (notification === "CAROUSEL_NEXT") {
            this.manualTransition(undefined, 1);
            this.restartTimer();
        }
        if (notification === "CAROUSEL_PREVIOUS") {
            this.manualTransition(undefined, -1);
            this.restartTimer();
        }
        if (notification === "CAROUSEL_GOTO") {
            if (typeof payload === "number" || typeof payload === "string") {
                try {
                    this.manualTransition(parseInt(payload) - 1);
                    this.restartTimer();
                } catch (err) {
                    console.warn("Could not navigate to slide " + payload);
                }
            } else if (typeof payload === "object") {
                try {
                    this.manualTransition(undefined, 0, payload.slide);
                    this.restartTimer();
                } catch (err) {
                    console.warn("Could not navigate to slide " + payload.slide);
                }
            }
        }
    },

    setUpTransitionTimers: function(positionIndex) {
        var modules, timer = this.config.transitionInterval;
        modules = MM.getModules().exceptModule(this).filter(function(module) {
            if (positionIndex === null) {
                return this.config.ignoreModules.indexOf(module.name) === -1;
            }
            return ((this.config[positionIndex].ignoreModules.indexOf(module.name) === -1) && (module.data.position === positionIndex));
        }, this);

        if (this.config.mode === 'slides') {
            modules.slides = this.config.slides;
        }

        if (positionIndex !== null) {
            if ((this.config[positionIndex].overrideTransitionInterval !== undefined) && (this.config[positionIndex].overrideTransitionInterval > 0)) {
                timer = this.config[positionIndex].overrideTransitionInterval;
            }
        }

        modules.currentIndex = -1;
        modules.showPageIndicators = this.config.showPageIndicators;
        modules.showPageControls = this.config.showPageControls;
        modules.showPageTitles = this.config.showPageTitles;
        modules.slideTransitionSpeed = this.config.slideTransitionSpeed;
        this.moduleTransition.call(modules);

        // Reference to function for manual transitions
        this.manualTransition = this.moduleTransition.bind(modules);

        if (this.config.mode !== "slides" || (this.config.mode === "slides" && timer > 0)) {
            // We set a timer to cause the page transitions
            // If we're in slides mode and the timer is set to 0, we only use manual transitions
            this.transitionTimer = setInterval(this.manualTransition, timer);
        }
    },

    moduleTransition: function(goToIndex = -1, goDirection = 0, goToSlide = undefined) {
        var i, resetCurrentIndex = this.length;
        if (this.slides !== undefined) {
            resetCurrentIndex = Object.keys(this.slides).length;
        }

        // Update the current index
        if (goToSlide) {
            let slide = Object.keys(this.slides).find((s, i) => {
                if (goToSlide === s) {
                    this.currentIndex = i;
                    return true;
                }
                return false;
            });
        } else if (goToIndex === -1) { // Go to a specific slide?
            if (goDirection === 0) {
                this.currentIndex += 1; // Normal Transition, Increment by 1
            } else {
                // console.log("Currently on slide " + this.currentIndex + " and going to slide " + (this.currentIndex + goDirection));
                this.currentIndex += goDirection; // Told to go a specific direction
            }
            if (this.currentIndex >= resetCurrentIndex) { // Wrap-around back to beginning
                this.currentIndex = 0;
            } else if (this.currentIndex < 0) {
                this.currentIndex = resetCurrentIndex - 1; // Went too far backwards, wrap-around to end
            }
        } else if (goToIndex >= 0 && goToIndex < resetCurrentIndex) {
            this.currentIndex = goToIndex; // Go to a specific slide if in range
        }

        /* selectWrapper(position)
         * Select the wrapper dom object for a specific position.
         *
         * argument position string - The name of the position.
         */
        var selectWrapper = function(position) {
            var classes = position.replace("_", " ");
            var parentWrapper = document.getElementsByClassName(classes);
            if (parentWrapper.length > 0) {
                var wrapper = parentWrapper[0].getElementsByClassName("container");
                if (wrapper.length > 0) {
                    return wrapper[0];
                }
            }
        };

        for (i = 0; i < this.length; i += 1) {
            // There is currently no easy way to discover whether a module is ALREADY shown/hidden
            // In testing, calling show/hide twice seems to cause no issues
            //console.log("Processing " + this[i].name);
            if ((this.slides === undefined) && (i === this.currentIndex)) {
                this[i].show(this.slideTransitionSpeed, { lockString: "mmmc" });
            } else if (this.slides !== undefined) {
                // Handle slides
                var mods = this.slides[Object.keys(this.slides)[this.currentIndex]];
                var show = false;
                // Loop through all of the modules that are supposed to be in this slide
                for (var s = 0; s < mods.length; s++) {
                    if (typeof mods[s] === "string" && mods[s] === this[i].name) {
                        // If only the module name is given as a string, and it matches, show the module
                        this[i].show(this.slideTransitionSpeed, { lockString: "mmmc" });
                        show = true;
                        break;
                    } else if (typeof mods[s] === "object" && ("name" in mods[s]) && mods[s].name === this[i].name) {
                        // If the slide definition has an object, and it's name matches the module continue
                        // check if carouselId is set (mutiple module instances) and this is not the one we should show
                        if ((typeof mods[s].carouselId !== "undefined") &&
                            (typeof this[i].data.config.carouselId !== "undefined") &&
                            (this[i].data.config.carouselId.toString().indexOf(mods[s].carouselId) === -1)) { break; }
                        if (typeof mods[s].classes === "string") {
                            // Check if we have any classes we're supposed to add
                            var dom = document.getElementById(this[i].identifier);
                            // Remove any classes added by this module (other slides)
                            dom.className = dom.className.split("mmmc")[0];
                            if (mods[s].classes) {
                                // check for an empty classes tag (required to remove classes added from other slides)
                                // If we have a valid class list, add the classes
                                dom.classList.add("mmmc");
                                dom.classList.add(mods[s].classes);
                            }
                        }

                        if (typeof mods[s].position === "string") {
                            // Check if we were given a position to change, if so, move the module to the new position
                            selectWrapper(mods[s].position).appendChild(document.getElementById(this[i].identifier));
                        }
                        // Finally show the module
                        this[i].show(this.slideTransitionSpeed, { lockString: "mmmc" });
                        show = true;
                        break;
                    }
                }
                // The module is not in this slide.
                if (!show) { this[i].hide(0, { lockString: "mmmc" }); }
            } else {
                // We aren't using slides and this module shouldn't be shown.
                this[i].hide(0, { lockString: "mmmc" });
            }
        }

        // Update the DOM if we're using it.
        if ((this.slides !== undefined) && (this.showPageIndicators || this.showPageControls || this.showPageTitles)) {
            var slider = document.getElementById("slider_" + this.currentIndex);
            slider.checked = true;
            var label;

            if (this.showPageIndicators) {
                var currPages = document.getElementsByClassName("MMMCarouselCurrentPage");
                if (currPages && currPages.length > 0) {
                    for (i = 0; i < currPages.length; i++) {
                        currPages[i].classList.remove('MMMCarouselCurrentPage');
                    }
                }
                document.getElementById("sliderLabel_" + this.currentIndex).classList.add('MMMCarouselCurrentPage');
            }

            if (this.showPageControls) {
                var currBtns = document.getElementsByClassName("MMMCarouselAvailable");
                if (currBtns && currBtns.length > 0) {
                    while (currBtns.length > 0) {
                        currBtns[0].classList.remove('MMMCarouselAvailable');
                    }
                }
                if (this.currentIndex !== resetCurrentIndex - 1) {
                    // console.log("Trying to enable button sliderNextBtn_" + (this.currentIndex+1));
                    document.getElementById("sliderNextBtn_" + (this.currentIndex + 1)).classList.add('MMMCarouselAvailable');
                }
                if (this.currentIndex !== 0) {
                    // console.log("Trying to enable button sliderPrevBtn_" + (this.currentIndex-1));
                    document.getElementById("sliderPrevBtn_" + (this.currentIndex - 1)).classList.add('MMMCarouselAvailable');
                }
            }

            if (this.showPageTitles) {
                var hdr = document.getElementById('MMMCarouselPageTitle');
                hdr.textContent = Object.keys(this.slides)[this.currentIndex];
            }
        }
    },

    restartTimer: function() {
        if (this.config.transitionTimer > 0) {
            // Restart the timer
            clearInterval(this.transitionTimer);
            this.transitionTimer = setInterval(this.manualTransition, this.config.transitionInterval);
        }
    },

    manualTransitionCallback: function(slideNum) {
        // console.log("manualTransition was called by slider_" + slideNum);
        // Perform the manual transitio
        this.manualTransition(slideNum);
        this.restartTimer();
    },

    getStyles: function() {
        return ["MMM-Carousel.css"];
    },

    /* getDom()
     * This method generates the dom which needs to be displayed. This method is called by the Magic Mirror core.
     * This method needs to be subclassed if the module wants to display info on the mirror.
     *
     * return domobject - The dom to display.
     */
    getDom: function() {
        var self = this;

        function makeOnChangeHandler(id) {
            return function() {
                self.manualTransitionCallback(id);
            };
        }

        if (this.config.mode === "slides" && (this.config.showPageIndicators ||
                                              this.config.showPageControls ||
                                              this.config.showPageTitles)) {

            var div = document.createElement("div");
            div.className = "MMMCarouselContainer";

            var paginationWrapper = document.createElement("div");
            paginationWrapper.className = "slider-pagination";

            for (var i = 0; i < Object.keys(this.config.slides).length; i++) {
                var input = document.createElement("input");
                input.type = "radio";
                input.name = "slider";
                input.id = "slider_" + i;
                input.className = "slide-radio";
                input.onchange = makeOnChangeHandler(i);
                paginationWrapper.appendChild(input);
            }

            if (this.config.showPageIndicators) {
                for (i = 0; i < Object.keys(this.config.slides).length; i++) {
                    var label = document.createElement("label");
                    label.setAttribute("for", "slider_" + i);
                    label.id = "sliderLabel_" + i;
                    paginationWrapper.appendChild(label);
                }
            }

            div.appendChild(paginationWrapper);

            if (this.config.showPageControls) {
                var nextWrapper = document.createElement("div");
                nextWrapper.className = "next control";

                var previousWrapper = document.createElement("div");
                previousWrapper.className = "previous control";

                for (var j = 0; j < Object.keys(this.config.slides).length; j++) {
                    if (j !== 0) {
                        var nCtrlLabelWrapper = document.createElement("label");
                        nCtrlLabelWrapper.setAttribute("for", "slider_" + j);
                        nCtrlLabelWrapper.id = "sliderNextBtn_" + j;
                        nCtrlLabelWrapper.innerHTML = '<i class="fa fa-arrow-circle-right"></i>';
                        nextWrapper.appendChild(nCtrlLabelWrapper);
                    }

                    if (j !== Object.keys(this.config.slides).length - 1) {
                        var pCtrlLabelWrapper = document.createElement("label");
                        pCtrlLabelWrapper.setAttribute("for", "slider_" + j);
                        pCtrlLabelWrapper.id = "sliderPrevBtn_" + j;
                        pCtrlLabelWrapper.innerHTML = '<i class="fa fa-arrow-circle-left"></i>';
                        previousWrapper.appendChild(pCtrlLabelWrapper);
                    }
                }

                div.appendChild(nextWrapper);
                div.appendChild(previousWrapper);
            }

            if (this.config.showPageTitles) {
                var hdr = document.createElement('header');
                hdr.setAttribute("id", "MMMCarouselPageTitle");
                hdr.classList.add('module-header');
                if (this.config.showPageTitles == 'top') {
                    var top_bar = document.getElementsByClassName('region top bar')[0];
                    top_bar.insertBefore(hdr, top_bar.firstChild);
                } else if (this.config.showPageTitles == 'bottom' ) {
                    hdr.style.cssText = 'margin: -10px;';
                    div.insertBefore(hdr, div.firstChild);
                }
            }
            return div;
        }
    },
});